Add Config.ofe, refactor optional dependency usage

Morfent opened the pull request:
`nonDefaultDependencies` doesn't seem to be a standard package.json field
anymore. This version of `ofe` fixes a crash when attempting to build the
package on Windows.

npm v5 complains about `package-lock.json` not existing.

Zarel commented on package.json:
```diff
@@ -10,10 +10,8 @@
     "nodemailer": "4.0.1",
-    "node-static": "0.7.9"
-  },
-  "nonDefaultDependencies": {
```
I don't know if it was ever a real thing? We just intentionally did that because we didn't want npm to install it automatically, which it does for optional dependencies.

Zarel replied:
ahem

Morfent replied:
Would it make more sense just to not include it as a dependency in package.json, then warn the user if it's not installed when it's called in `sockets.js` to run `npm install --no-save ofe` if it's desired? I don't really like the idea of using `nonDefaultDependencies` when it's not a valid `package.json` field

Zarel replied:
Perhaps mention how to install it in `config.js`... yeah, I can see that.

Zarel commented:
Also wtf is `package-lock.json`?

Morfent commented:
It seems to be new to npm v5

Zarel commented:
> One key detail about `package-lock.json` is that it cannot be published

meanwhile

> npm notice created a lockfile as package-lock.json. You should commit this file.

...I don't even know what it's for... I know that Facebook's `npm install` replacement also creates one, but I still have no idea why it needs to exist...

Zarel commented:
What does "cannot be published" even mean? I did find the detail that it does go in repositories, though.

Morfent commented:
> One key detail about package-lock.json is that it cannot be published

[npm-shrinkwrap.json](https://docs.npmjs.com/files/shrinkwrap.json) serves the same purpose as `package-lock.json`, but can be published. Its purpose is apparently to ensure each user running PS has the same versions of the same dependencies installed

Morfent commented:
What they mean by "cannot be published" is that it cannot be published to npm's registry. That's what `shrinkwrap.json` is meant for

Zarel commented:
Okay, I committed a `package-lock.json`. Shouldn't be in the same commit as these other changes, anyway.

Morfent commented:
Dropped the `package-lock.json` commit

Morfent commented:
Added a `Config.heapdumps` setting and added a warning to the user to install `ofe` if `Config.heapdumps` is set to true but it's not already installed

Zarel commented:
Should probably just call it `Config.ofe`.

We should probably also treat our other optional dependencies this way.

Morfent commented:
Which ones? `node-static`, `cloud-env`, and `nodemailer`?
Also, `http-proxy` is completely unused.

Morfent commented:
Optional dependencies can be left out by running `npm install --production --no-optional`. Maybe the launch script should install dependencies using that instead of `npm install --production`, so `ofe` and the rest of the optional dependencies can still be kept in `package.json`?

Morfent commented:
Not entirely sure if this is what you were asking for, but I wrote it anyway

Zarel commented:
`nodemailer` yes, `node-static` leave alone.

Zarel commented on pokemon-showdown:
```diff
@@ -19,7 +19,7 @@ try {
 } catch (e) {
 	console.log('Installing dependencies...');
-	child_process.execSync('npm install --production', {stdio: 'inherit'});
+	child_process.execSync('npm install --production --no-optional', {stdio: 'inherit'});
```
Disagree, `cloud-env` and `node-static` are important defaults.

Zarel replied:
I'm not entirely sure why they're in there at all...

Morfent replied:
Should they be regular dependencies?

Zarel replied:
They should be optional dependencies.

Zarel replied:
I still don't approve of the `--no-optional` approach.

Zarel replied:
`sockjs` is required

`node-static` and `cloud-env` are optional but should be installed by default

`ofe` and `nodemailer` are optional and should not be installed by default

Morfent commented:
Made `node-static` and `cloud-env` regular dependencies, removed `Config.nodestatic`

Zarel commented on config/config-example.js:
```diff
@@ -10,6 +10,17 @@ exports.port = 8000;
+// cloudenv - automatically configure port and bind address using cloud environment
+//   This should only be enabled if the server is running on a cloud-based host,
+//   such as c9, Heroku, or OpenShift.
+exports.cloudenv = false;
```
This should not be here; the entire point of cloudenv is to do this by default instead of needing manual configuration.

Zarel commented on crashlogger.js:
```diff
@@ -50,24 +49,34 @@ module.exports = function crashLogger(err, description, data = null) {
+				transport = require('nodemailer').createTransport(Config.crashguardemail.options);
+			} catch (e) {
+				if (e.code !== 'MODULE_NOT_FOUND') throw e; // Config.crashguardemail was likely misconfigured
+				throw new Error(
```
Um, is this going to lead to infinite recursion?

Morfent replied:
No. When the errors are thrown in the catch statement here, they don't cause any 'uncaughtException' event to be emitted by `process`, so the errors don't get passed back to `crashlogger` again. I'll still need to rewrite this part if `nodemailer` shouldn't be installed by default though

Morfent replied:
The server will still lockdown when it's supposed to if an error's thrown here, also

Morfent commented:
Made the changes you recommended

Zarel commented:
I'm mildly disappointed that we now have a config setting for ofe, instead of being able to do `npm install ofe` to enable it...

Zarel commented:
I guess it makes sense, though.
